kafka/consumer.go

```diff
@@ -424,30 +424,30 @@ func (c *Consumer) Close() (err error) {
 		close(c.events)
 	}
 
-	// librdkafka's rd_kafka_consumer_close() will block
-	// and trigger the rebalance_cb() if one is set, if not, which is the
-	// case with the Go client since it registers EVENTs rather than callbacks,
-	// librdkafka will shortcut the rebalance_cb and do a forced unassign.
-	// But we can't have that since the application might need the final RevokePartitions
-	// before shutting down. So we trigger an Unsubscribe() first, wait for that to
-	// propagate (in the Poll loop below), and then close the consumer.
-	c.Unsubscribe()
-
-	// Poll for rebalance events
-	for {
-		c.Poll(10 * 1000)
-		if int(C.rd_kafka_queue_length(c.handle.rkq)) == 0 {
-			break
+	doneChan := make(chan bool)
+	go func() {
+		C.rd_kafka_consumer_close(c.handle.rk)
+		// wake up Poll()
+		C.rd_kafka_queue_yield(c.handle.rkq)
+		doneChan <- true
+	}()
+
+	// wait for consumer_close() to finish while serving c.Poll() for rebalance callbacks/events
+	run := true
+	for run {
+		c.Poll(100)
+		select {
+		case <-doneChan:
```

Comment:
I think this will block until doneChan is written to, which means if Poll() is triggered once for some other reason, e.g., a stats callback, then it will hang here.

Reply:
Can we use `int(C.rd_kafka_queue_length(c.handle.rkq)) == 0` to exist the loop which is same with the previous way?
```
for run {
		c.Poll(100)
		select {
		case <-doneChan:
			run = false

		default:
			if int(C.rd_kafka_queue_length(c.handle.rkq)) == 0 {
				run = false
			}
		}
	}
```

Reply:
And I'm also confused why it will hang, I tested with this code with the existing change in this PR, it doesn't hang

```
func TestConsumerClose1(t *testing.T) {

	c, err := NewConsumer(&ConfigMap{
		"group.id":               "gotest"})
	if err != nil {
		t.Fatalf("%s", err)
	}

	startClose := time.Now()
	err = c.Close()
	if err != nil {
		panic(err)
	}
	fmt.Println(fmt.Sprintf("consumer closed. closing time: %v", time.Since(startClose)))
}
```

Reply:
I think with the following sequence of events it will hang:
 1. application calls Close()
 2. Close() is blocking on c.Poll()
 3. A stats event (or something else which is not a rebalance event) triggers the Poll and Poll() wakes up and returns
 4. Close() is now blocking on doneChan
 5. librdkafka emits a rebalance event and waits for the app/go to ack it
 6. .. but since Close() is no longer calling Poll() the rebalance event is never handled.
 7. indefinite hang

Reply:
I think this is easily solved by moving the Poll to the default case in the select, like in other places in the code.

Reply:
```
for run {
		select {
		case <-doneChan:
**-------------->// Is it possible that after the doneChan, there are still other events we need to handle? We always have to wait to read the doneChan value. If we exit from default, the doneChan will report some error**
			run = false

		default:
                         c.Poll(100)
			if int(C.rd_kafka_queue_length(c.handle.rkq)) == 0 {
				run = false
			}
		}
	}
```

Reply:
I don't think so, when doneChan is done we don't need to call Poll no more.

Reply:
Why is the queue_length() check needed?

Reply:
We don't need it, I already removed it from my code.